### PR TITLE
Refactored @BillXMA availability PR, added tests, and refactored camping.py

### PR DESCRIPTION
**Pull request by colossatr0n**
* Refactored specifying campsite IDs. Original implementation was done by BillXMA [here](https://github.com/banool/recreation-gov-campsite-checker/pull/35).
  * Added argument to output campsite IDs with their availability (--show-campsite-info):
```
$ python camping.py --start-date 2018-07-20 --end-date 2018-07-23 --parks 232448 232450 232447 232770 --show-campsite-info --nights 1 
There are campsites available from 2018-07-20 to 2018-07-23!!!
🏕 ELK CREEK CAMPGROUND (SAWTOOTH NF) (232042): 1 site(s) available out of 1 site(s)
  * Site 69800 is available on the following dates:
    * 2018-07-20 -> 2018-07-21 
    * 2018-07-21 -> 2018-07-22
``` 
  
  * Refactored camping.py:
    * Moved web requests into RecreationClient.
    * Converted constant strings to enums.
    * Moved utility functions into utility files.
    * Moved argparse logic into its own class.
  
  Added tests.

### CODE REVIEW

File: camping.py [hunk 1]
```diff
@@ -1,64 +1,32 @@
...
-
-INPUT_DATE_FORMAT = "%Y-%m-%d"
-ISO_DATE_FORMAT_REQUEST = "%Y-%m-%dT00:00:00.000Z"
-ISO_DATE_FORMAT_RESPONSE = "%Y-%m-%dT00:00:00Z"
-
-SUCCESS_EMOJI = "🏕"
-FAILURE_EMOJI = "❌"
-
-headers = {"User-Agent": UserAgent(verify_ssl=False).random}
-
```
Comment (colossatr0n) [inline camping.py:33]: This block was moved into RecreationClient, formatter.py, and emoji.py.

File: camping.py [hunk 2]
```diff
@@ -1,64 +1,32 @@
...
-    """
-    date_formatted = datetime.strftime(date_object, format_string)
-    return date_formatted
-
-
-def site_date_to_human_date(date_string):
-    date_object = datetime.strptime(date_string, ISO_DATE_FORMAT_RESPONSE)
-    return format_date(date_object, format_string=INPUT_DATE_FORMAT)
-
-
```
Comment (colossatr0n) [inline camping.py:48]: vv Moved into RecreationClient.

File: camping.py [hunk 3]
```diff
@@ -1,64 +1,32 @@
...
-    formats the date in the format that the API wants.
-    """
-    date_formatted = datetime.strftime(date_object, format_string)
-    return date_formatted
-
-
-def site_date_to_human_date(date_string):
-    date_object = datetime.strptime(date_string, ISO_DATE_FORMAT_RESPONSE)
-    return format_date(date_object, format_string=INPUT_DATE_FORMAT)
-
```
Comment (colossatr0n) [inline camping.py:47]: ^^ Moved into formatter.py.

File: camping.py [hunk 4]
```diff
@@ -101,37 +67,51 @@ def get_park_information(park_id, start_date, end_date, campsite_type=None):
...
+                    len(campsite_ids) > 0
+                    and int(campsite_data["campsite_id"]) not in campsite_ids
+                ):
                     continue
+
                 available.append(date)
             if available:
                 a += available
 
     return data
```
Comment (colossatr0n) [inline camping.py:94]: vv Moved into RecreationClient.

File: camping.py [hunk 5]
```diff
@@ -101,37 +67,51 @@ def get_park_information(park_id, start_date, end_date, campsite_type=None):
...
+                if (
+                    campsite_type
+                    and campsite_type != campsite_data["campsite_type"]
+                ):
+                    continue
+
+                if (
+                    len(campsite_ids) > 0
+                    and int(campsite_data["campsite_id"]) not in campsite_ids
+                ):
```
Comment (colossatr0n) [inline camping.py:85]: Used black and isort before pushing.

File: camping.py [hunk 6]
```diff
@@ -266,76 +281,15 @@ def main(parks, json_output=False):
...
-        help="Read list of park ID(s) from stdin instead",
-    )
-
+    parser = CampingArgParser()
     args = parser.parse_args()
 
     if args.debug:
         LOG.setLevel(logging.DEBUG)
 
-    parks = args.parks or [p.strip() for p in sys.stdin]
```
Comment (colossatr0n) [inline camping.py:335]: This is handled in CampingArgParser. It does all the validation and argument manipulation when the args are parsed.

File: clients/recreation_client.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+class RecreationClient:
+
+    BASE_URL = "https://www.recreation.gov"
+    AVAILABILITY_ENDPOINT = BASE_URL + "/api/camps/availability/campground/{park_id}/month"
+    MAIN_PAGE_ENDPOINT = BASE_URL + "/api/camps/campgrounds/{park_id}"
+
+    headers = {"User-Agent": UserAgent(verify_ssl=False).random}
+
+    @staticmethod
```
Comment (banool) [inline clients/recreation_client.py:19]: Let's make this a classmethod instead and use `cls.AVAILABILITY_ENDPOINT`, `cls._send_request`, etc.

File: utils/camping_argparser.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import argparse
+import logging
+import sys
+
+from datetime import datetime
+
+from enums.date_format import DateFormat
+
+
+class CampingArgParser(argparse.ArgumentParser):
```
Comment (banool) [inline utils/camping_argparser.py:10]: s/ArgParser/ArgumentParser/

File: utils/camping_argparser.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            help="Read list of park ID(s) from stdin instead",
+        )
+
+    def parse_args(self, args=None, namespace=None):
+        args = super().parse_args(args, namespace)
+        args.parks = args.parks or [p.strip() for p in sys.stdin]
+        self._validate_args(args)
+        return args
+
+    @staticmethod
```
Comment (banool) [inline utils/camping_argparser.py:75]: Same comment as before, and could you apply it generally? If you have a method of a class that refers to the class, make it a class method.

File: tests/camping_test.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class CampingTest(unittest.TestCase):
+
+    def testAggregatesDataForMultipleCampsites(self):
+        park_info = {'1': [],
+                     '2': ['2022-06-22T00:00:00Z', '2022-06-23T00:00:00Z', '2022-06-26T00:00:00Z'],
+                     '3': ['2022-06-22T00:00:00Z', '2022-06-23T00:00:00Z', '2022-06-26T00:00:00Z',
+                               '2022-06-27T00:00:00Z', '2022-06-29T00:00:00Z']}
+        _, _, available_dates_by_campsite_id = \
```
Comment (banool) [inline tests/camping_test.py:13]: Does black actually do this, or did you add the backslash yourself?

File: camping.py [hunk 11]
```diff
@@ -1,64 +1,32 @@
...
 import logging
 import sys
+from collections import defaultdict
 from datetime import datetime, timedelta
-from dateutil import rrule
 from itertools import count, groupby
 
-import requests
-from fake_useragent import UserAgent
+from dateutil import rrule
```
Comment (banool) [inline camping.py:10]: Can you merge this import into line 7?

File: camping.py [hunk 12]
```diff
@@ -266,76 +281,15 @@ def main(parks, json_output=False):
...
         LOG.setLevel(logging.DEBUG)
 
-    parks = args.parks or [p.strip() for p in sys.stdin]
-
     try:
-        code = 0 if main(parks, json_output=args.json_output) else 1
+        code = 0 if main(args.parks, json_output=args.json_output) else 1
         sys.exit(code)
     except Exception:
         print("Something went wrong")
```
Comment (banool) [inline camping.py:295]: While we're here just remove this try / except. I'd rather see the exception than have it swallowed like this.

**Review by banool — CHANGES_REQUESTED** [review]
Thanks a lot for all of these changes, this is a much needed quality of life improvement. I don't have any major changes I'd like to see, just some nits.

Have you verified that the output without the new flags is the same and still works with the notifier?

File: tests/camping_test.py (hunk 10 above)
Comment (colossatr0n) [reply tests/camping_test.py:13]: I believe I added this one.

File: utils/camping_argparser.py (hunk 8 above)
Comment (colossatr0n) [reply utils/camping_argparser.py:10]: Can you explain a little more on this comment?

File: clients/recreation_client.py (hunk 7 above)
Comment (colossatr0n) [reply clients/recreation_client.py:19]: Done.

File: utils/camping_argparser.py (hunk 9 above)
Comment (colossatr0n) [reply utils/camping_argparser.py:75]: Done.

File: camping.py (hunk 11 above)
Comment (colossatr0n) [reply camping.py:10]: I don't believe so. `rrule` is from `dateutil`, not `datetime`.

File: camping.py (hunk 12 above)
Comment (colossatr0n) [reply camping.py:295]: Done.

File: camping.py (hunk 11 above)
Comment (banool) [reply camping.py:10]: Oh my bad, misread, thanks!

File: utils/camping_argparser.py (hunk 8 above)
Comment (banool) [reply utils/camping_argparser.py:10]: Just a nit, as in rename ArgParser to ArgumentParser. My bad, sed terminology 😛

File: tests/camping_test.py (hunk 10 above)
Comment (banool) [reply tests/camping_test.py:13]: Could we remove it and let black do its thing instead?

**Comment by colossatr0n** [comment]
> Thanks a lot for all of these changes, this is a much needed quality of life improvement. I don't have any major changes I'd like to see, just some nits.

No problem! Thanks for maintaining this repo.

> Have you verified that the output without the new flags is the same and still works with the notifier?

Let me double check.

File: tests/camping_test.py (hunk 10 above)
Comment (colossatr0n) [reply tests/camping_test.py:13]: Yeah! Will remove. I'll rerun black one more time as well just in case anything else got missed.

**Comment by banool** [comment]
Thanks for checking. I'd love to make the notifier take some structured input, json at the least, to make changes to the main script less likely to break it. As it is now, it does very janky parsing of the human readable output. But for now we have to be careful.

File: utils/camping_argparser.py (hunk 8 above)
Comment (colossatr0n) [reply utils/camping_argparser.py:9]: Oh hahah. I see now.

**Comment by colossatr0n** [comment]
Output without the new flags is the same and the notifier still works.

**Comment by banool** [comment]
Perfect. Rebase this PR since there are conflicts and then we'll be good to go.

**Comment by colossatr0n** [comment]
Rebased 👍

**Comment by banool** [comment]
Thanks for working on all these changes!!

**Comment by colossatr0n** [comment]
You're welcome! Thanks for accepting the PR.

**Pull request merged**